### PR TITLE
(PUP-7042) Mark strings in util

### PR DESCRIPTION
**Pull request by Magisus**
This commit marks user-facing error and info level strings in
`lib/puppet/util.rb` and `lib/puppet/util/*` for translation.

### CODE REVIEW

**Comment by Magisus** [comment]
Closing and reopening to fix Travis.

**Comment by Magisus** [comment]
Closing and reopening to fix CI.

File: lib/puppet/util.rb [hunk 1]
```diff
@@ -254,10 +256,10 @@ def which(bin)
           if e.to_s =~ /HOME/ and (Puppet::Util.get_env('HOME').nil? || Puppet::Util.get_env('HOME') == "")
             # if we get here they have a tilde in their PATH.  We'll issue a single warning about this and then
             # ignore this path element and carry on with our lives.
-            Puppet::Util::Warnings.warnonce("PATH contains a ~ character, and HOME is not set; ignoring PATH element '#{dir}'.")
+            Puppet::Util::Warnings.warnonce(_("PATH contains a ~ character, and HOME is not set; ignoring PATH element '%{dir}'.") % { dir: dir })
```
Comment (MosesMendoza) [inline lib/puppet/util.rb]: Can you add a translator note re: PATH, ~, and HOME being references to variable names and not intended for translation?

File: lib/puppet/util.rb [hunk 2]
```diff
@@ -254,10 +256,10 @@ def which(bin)
...
           elsif e.to_s =~ /doesn't exist|can't find user/
             # ...otherwise, we just skip the non-existent entry, and do nothing.
-            Puppet::Util::Warnings.warnonce("Couldn't expand PATH containing a ~ character; ignoring PATH element '#{dir}'.")
+            Puppet::Util::Warnings.warnonce(_("Couldn't expand PATH containing a ~ character; ignoring PATH element '%{dir}'.") % { dir: dir })
```
Comment (MosesMendoza) [inline lib/puppet/util.rb]: Same as above

File: lib/puppet/util/command_line/trollop.rb [hunk 3]
```diff
@@ -146,7 +146,7 @@ def initialize *a, &b
   ## value, you must specify +:type+ as well.
 
   def opt name, desc="", opts={}
-    raise ArgumentError, "you already have an argument named '#{name}'" if @specs.member? name
+    raise ArgumentError, _("you already have an argument named '#{name}'") if @specs.member? name
```
Comment (MosesMendoza) [inline lib/puppet/util/command_line/trollop.rb]: Looks like `trollop` was missed on the variable interpolation fixes sweep.

File: lib/puppet/util/errors.rb [hunk 4]
```diff
@@ -37,11 +37,11 @@ def adderrorcontext(error, other = nil)
   # @return [String] description of file and line
   def error_context
     if file and line
-      " at #{file}:#{line}"
+      _(" at %{file}:%{line}") % { file: file, line: line }
```
Comment (MosesMendoza) [inline lib/puppet/util/errors.rb]: 😞  the way we did this, there's no way this is going to come through in some languages. sigh

Comment (Magisus) [inline lib/puppet/util/errors.rb]: I know, I hate these little "build up the error message" blocks. I guess we can see what the translator makes of it? Or leave it out entirely...

File: lib/puppet/util/ldap/connection.rb [hunk 5]
```diff
@@ -66,6 +66,6 @@ def start
       @connection.set_option(LDAP::LDAP_OPT_REFERRALS, LDAP::LDAP_OPT_ON)
       @connection.simple_bind(user, password)
   rescue => detail
-      raise Puppet::Error, "Could not connect to LDAP: #{detail}", detail.backtrace
+      raise Puppet::Error, _("Could not connect to LDAP: #{detail}"), detail.backtrace
```
Comment (MosesMendoza) [inline lib/puppet/util/ldap/connection.rb]: Missed variable interpolation

File: lib/puppet/util/ldap/manager.rb [hunk 6]
```diff
@@ -213,14 +213,14 @@ def search(sfilter = nil)
   # Update the ldap entry with the desired state.
   def update(name, is, should)
     if should[:ensure] == :absent
-      Puppet.info "Removing #{dn(name)} from ldap"
+      Puppet.info _("Removing #{dn(name)} from ldap")
```
Comment (MosesMendoza) [inline lib/puppet/util/ldap/manager.rb]: Missed variable interpolation

File: lib/puppet/util/ldap/manager.rb [hunk 7]
```diff
@@ -213,14 +213,14 @@ def search(sfilter = nil)
...
       delete(name)
       return
     end
 
     # We're creating a new entry
     if is.empty? or is[:ensure] == :absent
-      Puppet.info "Creating #{dn(name)} in ldap"
+      Puppet.info _("Creating #{dn(name)} in ldap")
```
Comment (MosesMendoza) [inline lib/puppet/util/ldap/manager.rb]: Same

File: lib/puppet/util/log/destinations.rb [hunk 8]
```diff
@@ -18,7 +18,7 @@ def initialize
     begin
       facility = Syslog.const_get("LOG_#{str.upcase}")
     rescue NameError
-      raise Puppet::Error, "Invalid syslog facility #{str}", $!.backtrace
+      raise Puppet::Error, _("Invalid syslog facility #{str}"), $!.backtrace
```
Comment (MosesMendoza) [inline lib/puppet/util/log/destinations.rb]: Missing interpolation fixes in this file

File: lib/puppet/util/ldap/connection.rb (hunk 5 above)
Comment (Magisus) [inline lib/puppet/util/ldap/connection.rb]: I think I may need to redo it for the `util` directory. I think I used a different invocation that caused some stuff to get missed, due to there being a couple of non-Ruby files in this directory.

File: lib/puppet/util/logging.rb [hunk 9]
```diff
@@ -281,8 +281,8 @@ def issue_deprecation_warning(message, key, file, line, use_caller)
         $deprecation_warnings[key] = message
         call_trace = use_caller ?
           (offender || get_deprecation_offender).join('; ') :
-          "#{file || 'unknown'}:#{line || 'unknown'}"
-        warning("#{message}\n   (at #{call_trace})")
+          "#{file || _('unknown')}:#{line || _('unknown')}"
```
Comment (MosesMendoza) [inline lib/puppet/util/logging.rb]: same re: interpolation of `file` and `line`, but I think you have to move the conditional into the parameters hash, ie

_("%{file}:%{line}") % { file: file || 'unknown', line: line || 'unknown' }

Comment (MosesMendoza) [inline lib/puppet/util/logging.rb]: which kind of begs the question - how would 'unknown' get translated? maybe define an unknown string variable first?

ie.
unknown = _("unknown")
_("%{file}:%{line}") % { file: file || unknown, line: line || unknown }

File: lib/puppet/util/network_device/cisco/device.rb [hunk 10]
```diff
@@ -51,7 +51,7 @@ def execute(cmd)
       if out =~ /^%/mo or out =~ /^Command rejected:/mo
         # strip off the command just sent
         error = out.sub(cmd,'')
-        Puppet.err "Error while executing '#{cmd}', device returned: #{error}"
+        Puppet.err _("Error while executing '#{cmd}', device returned: #{error}")
```
Comment (MosesMendoza) [inline lib/puppet/util/network_device/cisco/device.rb]: Missing variable interpolation fixes in this file

File: lib/puppet/util/network_device/cisco/interface.rb [hunk 11]
```diff
@@ -87,7 +87,7 @@ def command(command)
       if out =~ /^%/mo or out =~ /^Command rejected:/mo
         # strip off the command just sent
         error = out.sub(command,'')
-        Puppet.err "Error while executing '#{command}', device returned: #{error}"
+        Puppet.err _("Error while executing '#{command}', device returned: #{error}")
```
Comment (MosesMendoza) [inline lib/puppet/util/network_device/cisco/interface.rb]: same

File: lib/puppet/util/network_device/config.rb [hunk 12]
```diff
@@ -52,7 +52,7 @@ def parse
           when /^\[([\w.-]+)\]\s*$/ # [device.fqdn]
             name = $1
             name.chomp!
-            raise Puppet::Error, "Duplicate device found at line #{count}, already found at #{device.line}" if devices.include?(name)
+            raise Puppet::Error, _("Duplicate device found at line #{count}, already found at #{device.line}") if devices.include?(name)
```
Comment (MosesMendoza) [inline lib/puppet/util/network_device/config.rb]: Same

File: lib/puppet/util/network_device/transport/ssh.rb [hunk 13]
```diff
@@ -33,21 +33,21 @@ def connect(&block)
       Puppet.debug("connecting to #{host} as #{user}")
       @ssh = Net::SSH.start(host, user, :port => port, :password => password, :timeout => timeout)
     rescue TimeoutError
-      raise TimeoutError, "timed out while opening an ssh connection to the host", $!.backtrace
+      raise TimeoutError, _("timed out while opening an ssh connection to the host"), $!.backtrace
     rescue Net::SSH::AuthenticationFailed
-      raise Puppet::Error, "SSH authentication failure connecting to #{host} as #{user}", $!.backtrace
+      raise Puppet::Error, _("SSH authentication failure connecting to #{host} as #{user}"), $!.backtrace
```
Comment (MosesMendoza) [inline lib/puppet/util/network_device/transport/ssh.rb]: Same

File: lib/puppet/util/logging.rb (hunk 9 above)
Comment (Magisus) [inline lib/puppet/util/logging.rb]: Sounds good. Only the "unknown" string and final message at the bottom need to be extracted then. The middle man is all formatted variables. Updated.

File: lib/puppet/util/windows/principal.rb [hunk 14]
```diff
@@ -93,7 +93,7 @@ def self.lookup_account_name(system_name = nil, account_name)
     def self.lookup_account_sid(system_name = nil, sid_bytes)
       system_name_ptr = FFI::Pointer::NULL
       if (sid_bytes.nil? || (!sid_bytes.is_a? Array) || (sid_bytes.length == 0))
-        raise Puppet::Util::Windows::Error.new('Byte array for lookup_account_sid must not be nil and must be at least 1 byte long')
+        raise Puppet::Util::Windows::Error.new(_('Byte array for lookup_account_sid must not be nil and must be at least 1 byte long'))
```
Comment (MosesMendoza) [inline lib/puppet/util/windows/principal.rb]: Can you leave a translator note that `lookup_account_sid` is a variable name?

**Pull request merged**